### PR TITLE
Add `hex` crate

### PR DESCRIPTION
**Pull request by tcharding**
Draft until #1230 gets resolved (and lock files added to this PR). Also until @Kixunil is ok with it. Apart from that it should be good to go.

Add a new `hex` crate made up of formatting code from `internals::hex` and parsing code from `hashes::hex`.

I've tried to make this easy to review, lets keep further API improvements separate from the task of this PR which is separating the hex stuff out once and for all so we can get on with shit!

## Note on reviewing

I intentionally put the "import code" patches separate so during review one can checkout the respective commit and run a `diff` command to see the changes

For commit `9f464bdb Import code from internals`
```
- git checkout 9f464bdb
- diff internals/src/display.rs hex/src/display.rs
- diff internals/src/buff_encoder.rs hex/src/buff_ecoder.rs
```
And for commit `6358a193 Import hex code from hashes`
```
- git checkout 6358a193
- diff hashes/src/hex.rs hex/src/parse.rs
```
### TODO in later PRs

- Add benchmarks
- API improvements
- Stabalize and get to 1.0 - this is as good a crate as any to work out exactly how we get our crates to v1.0

### CODE REVIEW

**Comment by Kixunil** [comment]
I suggest that we do 3 by simply importing the crate into `internals`. At least temporarily.

>  lets keep further API improvements separate from the task

OK, good that you wrote that :D

File: hex/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+[package]
+name = "fast-hex"
```
Comment (Kixunil) [inline hex/Cargo.toml:2]: Nice name, but we should also uphold it when parsing. :) Of course we will.

File: hex/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+version = "0.1.0"
+authors = ["Andrew Poelstra <apoelstra@wpsoftware.net>"]
+license = "CC0-1.0"
+description = "Fast hex encoding and decoding"
+homepage = "https://github.com/rust-bitcoin/hex/"
+repository = "https://github.com/rust-bitcoin/"
```
Comment (Kixunil) [inline hex/Cargo.toml:8]: `homepage` shouldn't point to the same URL as `repository` and `repository` shouldn't point to an organization.

File: hex/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+documentation = "https://docs.rs/fast-hex/"
+keywords = ["encoding", "hex", "hexadecimal"]
```
Comment (Kixunil) [inline hex/Cargo.toml:10]: `performance`?

File: hex/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+readme = "README.md"
+edition = "2018"
+
+[features]
+default = ["std"]
+std = ["alloc"]
+# If you disable std, you can still use Read/Write via the "io" feature.
+io = ["core2"]
```
Comment (Kixunil) [inline hex/Cargo.toml:18]: Why rename the feature? We shouldn't need it, just use `core2`.

File: hex/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+# If you disable std, you can still use Vec/String via the "alloc" feature.
+alloc = []
+
+[package.metadata.docs.rs]
+features = ["std", "alloc", "io"]
+rustdoc-args = ["--cfg", "docsrs"]
+
+[dependencies]
+# Only enable this if you explicitly do not want to use an allocator, otherwise enable "alloc".
```
Comment (Kixunil) [inline hex/Cargo.toml:27]: This comment looks out of place.

File: hex/README.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# Fast Hexadecimal Encoding/Decoding
+
+This is a hex library designed to be fast.
```
Comment (Kixunil) [inline hex/README.md:3]: "and have a conservative MSRV"

File: hex/README.md [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+## Minimum Supported Rust Version (MSRV)
+
+This library should always compile with any combination of features on **Rust 1.41.1**.
```
Comment (Kixunil) [inline hex/README.md:7]: "always" seems too strong. I definitely want to support fixed-sized arrays at some point with higher MSRV.

**Comment by Kixunil** [comment]
One more thing I'm not sure about is if it's productive to move this stuff before we have a nice API. Maybe we should warn that the crate is WIP?

File: hex/Cargo.toml (hunk 2 above)
Comment (tcharding) [reply hex/Cargo.toml:8]: Damn, its hard to get good help huh :)

File: hex/Cargo.toml (hunk 4 above)
Comment (tcharding) [reply hex/Cargo.toml:18]: I thought for downstream users "io" might make it easier to understand whats going on
```
rust-bitcoin = { version = "..", default-features = false, features = ["alloc", "io"]
```
Is obvious with having to lookup what the "io" feature is, as opposed to "core2". Although it's maybe only clear because I wrote it, perhaps "io" does not mean `Read` and `Write` _only_ (which is how we are using `core2`)?

File: hex/Cargo.toml (hunk 5 above)
Comment (tcharding) [reply hex/Cargo.toml:27]: Ah thanks, too many PRs and too much rebasing yesterday.

File: hex/README.md (hunk 6 above)
Comment (tcharding) [reply hex/README.md:3]: Done!

File: hex/README.md (hunk 7 above)
Comment (tcharding) [reply hex/README.md:7]: Perhaps MSRV does not imply `--all-features` (or all combinations) but it should definitely always compile with some subset of features, right?

If this sentence is wrong it needs fixing organisation wide.

**Comment by tcharding** [comment]
> One more thing I'm not sure about is if it's productive to move this stuff before we have a nice API. Maybe we should warn that the crate is WIP?

Oh I definitely think it is, this is the bottle neck for crate smashing AFAICT.

File: hex/Cargo.toml (hunk 4 above)
Comment (Kixunil) [reply hex/Cargo.toml:17]: Well, it goes against what is considered idiomatic. I don't find `io` particularly clear because it may suggest that the crate will be doing some I/O while it could be that it only uses in-memory operations. While `core2` is now required for some crates, we should really make our own traits and get rid of it.

File: hex/README.md (hunk 7 above)
Comment (Kixunil) [reply hex/README.md:7]: Hmm, I guess I'm misinterpreting "always" as "forever". But yes, we should have exception for some features.

**Comment by Kixunil** [comment]
> this is the bottle neck for crate smashing AFAICT.

We could just have it in `internals` and only move it later.

File: hex/Cargo.toml (hunk 2 above)
Comment (jeandudey) [reply hex/Cargo.toml]: I disagree with putting marketing words into crate descriptions, I mean it could be fast (and hope it is) but maybe not for every situation :).

Makes life easier when packaging for distributions that have guidelines on that sort of stuff, its hard to define what it's _fast_.

File: hex/Cargo.toml (hunk 3 above)
Comment (jeandudey) [reply hex/Cargo.toml]: See https://github.com/rust-bitcoin/rust-bitcoin/pull/1573#discussion_r1084367959

File: hex/Cargo.toml (hunk 4 above)
Comment (tcharding) [reply hex/Cargo.toml:17]: Ok, that's a strong enough argument for me. Will remove it.

**Comment by tcharding** [comment]
> We could just have it in internals and only move it later.

Oh boy, feels like we are going round in circles. Downstream users want to use our hex code I don't see why we don't just provide it for them right away - release early release often. `internals`, to my understanding, is for stuff internal that we don't expect downstream to use, please correct me if I'm wrong on the intention of `internals`.

File: hex/Cargo.toml (hunk 2 above)
Comment (tcharding) [reply hex/Cargo.toml]: Fair points, thanks man. I stayed away from `bitcoin-hex` because there is nothing in it that is specific to Bitcoin. The additional identifier is purely because `hex` is taken already. Will remove 'fast' though and fall back to `bitcoin-hex` unless we can come up with something better.

Comment (jeandudey) [reply hex/Cargo.toml]: Sorry didn't see the discussion on #1568 about other downstream crates needing a more up to date `hex` crate in which it would make sense to have a broader name. `bitcoin-hex` seems fine unless there's a better one.

Comment (apoelstra) [reply hex/Cargo.toml]: I also like `bitcoin-hex`. "What it has to do with bitcoin" is that it's maintained by the rust-bitcoin community, who has certain values about MSRV stability and minimizing dependency trees that others don't hold as strongly.

Comment (tcharding) [reply hex/Cargo.toml]: Convincing enough for me :) `bitcoin-hex` it is.

**Comment by tcharding** [comment]
I'm missing something that is making the no-std embedded test fail wildly. Not sure what it is?

File: hex/README.md (hunk 7 above)
Comment (tcharding) [reply hex/README.md:7]: Flagging this thread because I'm not 100% sure current (soon to be pushed) implementation resolves it.

**Comment by tcharding** [comment]
Lol, your sad face emoji sums up exactly how I felt yesterday afternoon when I saw this embedded fail.

**Comment by Kixunil** [comment]
I think this isn't as important as other things and could be done shortly after release because it's a separate crate. Let's focus on `bitcoin` 0.30 and do this afterwards?

**Comment by tcharding** [comment]
To release 0.30.0 we have to release `hashes` 0.13.0 and `internals`, if we do not create `hex` crate we either have to leave hex stuff in `hashes` 0.13 or we have to pursue your other idea to move it to `internals`. I personally would like to see `hashes` 0.13 have no hex module in it. I'm not convince that putting it in `internals` is a good idea also I"m afraid.

Do you still have some concerns over the `hex` crate in general?

**Comment by Kixunil** [comment]
Why it'd be a bad idea to move it into internals while `hex` is being worked on?

> Do you still have some concerns over the `hex` crate in general?

I do, I don't want to release half-baked crap as a separate crate. I didn't even get to review the API and think about it properly. I could but it'll just take a bunch of time which we could spend on releasing `bitcoin`&co sooner.

**Comment by tcharding** [comment]
> Why it'd be a bad idea to move it into internals while `hex` is being worked on?

I don't think we should encourage people to use `internals` as a dependency, it defeats IMO the whole point of having "internal" code, might as well have just left it in the main bitcoin crate. We have users that have explicitly stated that they want to use the hex code we write, if we put it in `internals` they will use an `internals` dependency.

> > Do you still have some concerns over the `hex` crate in general?
> 
> I do, I don't want to release half-baked crap as a separate crate. I didn't even get to review the API and think about it properly. 

The decoding API (`FromHex`) has been released in the main crate for ages (years?). The encoding API was written by you. 

> I could but it'll just take a bunch of time which we could spend on releasing `bitcoin`&co sooner.

I thought the idea was that the quickest way towards v1.0 / stable crates was to make little crates and then polish them individually.

**Comment by Kixunil** [comment]
>  if we put it in `internals` they will use an `internals` dependency.

I didn't see anyone threatening us to do that and I don't think we should cave in to such threats. We would simply have a brief period where we officially don't provide hex API at all. People who want to use our hex code can simply wait with dependency upgrade until it's ready.

> The decoding API (`FromHex`) has been released in the main crate for ages

That doesn't mean it's great. I have some ideas about error handling and looking into making it faster. (Mainly so that calling the crate `fast-hex` makes sense.)

> The encoding API was written by you.

Yep and I still feel like some things are missing. Also want to use const generics to handle arrays eventually.

>  the quickest way towards v1.0 / stable crates was to make little crates and then polish them individually?

Yes, that's quickest towards 1.0 but not 0.30 :)

**Comment by apoelstra** [comment]
>I didn't see anyone threatening us to do that and I don't think we should cave in to such threats

I don't consider it a threat -- but for my part I probably will depend on `internals` in my non-rust-bitcoin projects if that's what I need to get my own hex code. (Unsure. I may just depend on an old `bitcoin_hashes` as you suggest, which would probably be smaller.)

>That doesn't mean it's great.

Well, this is what 0.x versions are supposed to incidate. Would you be okay if we released it with version 0.0.1?

**Comment by mxz42** [comment]
Hi, just following along here. 

@tcharding 
> We have users that have explicitly stated that they want to use the hex code we write, if we put it in internals they will use an internals dependency.

Yes, fedimint belongs to that user group [as mentioned in the discussion page](https://github.com/rust-bitcoin/rust-bitcoin/discussions/1568#discussioncomment-4742934).

@apoelstra 
>  (Unsure. I may just depend on an old bitcoin_hashes as you suggest, which would probably be smaller.)

That's likely what we would do, too. At least as long as there no "official bitcoin hex crate". Or if the `internals` code improved by a lot so that it's worth switching over and depending on `internals`.

However, it would be nice to see a hex crate here that we could depend on going forward. 

Also, thanks @tcharding for putting this together

**Comment by Kixunil** [comment]
OK, releasing as 0.0.1 sounds interesting but I'd prefer to do this among the last steps near the release so that we could release sooner than this is ready if we wanted. Even if we released sooner I'd totally support prioritizing `hex`. (Note that if we go through `internals` we would only need to update `internals` to use `hex`.)

**Comment by tcharding** [comment]
The colloquial expression for this hex crate vs internals discussion is, I believe, "don't let perfect be the enemy of good". I fully agree we definitely can improve but I think what we have is good already.

**Comment by tcharding** [comment]
Rebased, used version "0.0.1"

**Comment by tcharding** [comment]
Rebased and checked the `hashes/src/hex.rs` import to `hex/src/parse.rs` was correct still after rebase.

File: hex/README.md (hunk 7 above)
Comment (apoelstra) [reply hex/README.md:7]: I think we should drop the word "always" if it makes Kix nervous. It's silly to claim any permanent rules for a 0.0.1 crate.

File: hex/src/lib.rs [hunk 8]
```diff
@@ -35,9 +35,9 @@ pub mod buf_encoder;
 pub mod display;
 pub mod parse;
 
-pub use buf_encoder::BufEncoder;
-pub use display::DisplayHex;
-pub use parse::{Error, FromHex};
+pub use buf_encoder::*;
+pub use display::*;
+pub use parse::*;
```
Comment (apoelstra) [inline hex/src/lib.rs:40]: In 4baf4410551693e5b0e210744abe69fe2bee1cc5:

I really dislike this change. It makes it impossible to tell where any hex types come from.

Other than that, ACK this.

File: hex/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,30 @@
+[package]
+name = "bitcoin-hex"
```
Comment (Kixunil) [inline hex/Cargo.toml:2]: Why the change? I find it really strange to name a general-purpose crate like this..

File: hex/examples/hexy.rs [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+impl fmt::LowerHex for Hexy {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::LowerHex::fmt(&self.as_bytes().as_hex(), f)
+    }
+}
+
+impl fmt::UpperHex for Hexy {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::UpperHex::fmt(&self.as_bytes().as_hex(), f)
```
Comment (Kixunil) [inline hex/examples/hexy.rs:58]: We should suggest the macro for fixed-sized cases since it's more performant.

File: bitcoin/Cargo.toml [hunk 11]
```diff
@@ -26,15 +26,16 @@ bitcoinconsensus-std = ["bitcoinconsensus/std", "std"]
...
+std = ["secp256k1/std", "bitcoin_hashes/std", "bech32/std", "bitcoin-internals/std", "bitcoin-hex/std"]
+no-std = ["core2", "bitcoin_hashes/alloc", "bitcoin_hashes/core2", "secp256k1/alloc", "bitcoin-hex/alloc", "bitcoin-hex/core2"]
 
 [package.metadata.docs.rs]
 all-features = true
 rustdoc-args = ["--cfg", "docsrs"]
 
 [dependencies]
 bitcoin-internals = { path = "../internals" }
+bitcoin-hex = { path = "../hex", default-features = false }
```
Comment (Kixunil) [inline bitcoin/Cargo.toml:38]: Consider using `hex = { package = "bitcoin-hex" ... }` to make access easier. (Like I did in hashes with internals, was thinking about doing it above too.)

Also I guess sooner or later we will have internals depend on hex.

**Review by Kixunil — COMMENTED** [review]
Just a few random observations. I don't feel well, so can't review properly.

File: hex/src/lib.rs (hunk 8 above)
Comment (tcharding) [reply hex/src/lib.rs:40]: Just to clarify, you mean you dislike the wildcard, right? The re-export is ok?

File: bitcoin/Cargo.toml (hunk 11 above)
Comment (tcharding) [reply bitcoin/Cargo.toml:38]: I used `pub extern crate bitcoin_hex as hex;` in `bitcoin/src/lib.rs` - achieves the same goal, I couldn't work out what was different so I went with the extern because we already have it for `hashes`.

Comment (tcharding) [reply bitcoin/Cargo.toml:38]: Started a discussion on something I've been wondering about for a while that relates to this: https://github.com/rust-bitcoin/rust-bitcoin/discussions/1661

File: hex/examples/hexy.rs (hunk 10 above)
Comment (tcharding) [reply hex/examples/hexy.rs:58]: Awesome, done!

File: hex/Cargo.toml (hunk 2 above)
Comment (tcharding) [reply hex/Cargo.toml]: @Kixunil, this is the thread of discussion that led me to change the crate name to `bitcoin-hex`.

File: hex/Cargo.toml (hunk 9 above)
Comment (tcharding) [reply hex/Cargo.toml:2]: Discussed above, I commented on the thread above to point you at it. I thought from that thread that we had consensus, will wait for more input and can rename if required.

**Comment by tcharding** [comment]
Force push is improves the example as suggested by Kix.

File: hex/src/lib.rs (hunk 8 above)
Comment (Kixunil) [reply hex/src/lib.rs:40]: I also question re-export of `buf_encoder`. The `display::*` items should be mostly sufficient. We should probably document that things inside `buf_encoder` are low-level details.

File: hex/Cargo.toml (hunk 2 above)
Comment (Kixunil) [reply hex/Cargo.toml]: How about something like `conservative-hex`/`solid-hex`...? Will pull GPT tomorrow to suggest more names if nobody beats me to it.

File: hex/README.md (hunk 7 above)
Comment (tcharding) [reply hex/README.md:7]: I used:
```

File: hex/src/lib.rs (hunk 8 above)
Comment (tcharding) [reply hex/src/lib.rs:40]: Added to the top of `buf_encoder.rs`
```
//! This is a low-level module, most uses should be satisfied by the `display` module instead.
```

Comment (tcharding) [reply hex/src/lib.rs:40]: And remove the re-export of `buf_encoder` stuff.

Comment (tcharding) [reply hex/src/lib.rs]: hex/src/lib.rs is now:
```rust
pub use display::{DisplayHex, DisplayArray};
pub use parse::{Error, FromHex, HexIterator};
```

**Comment by tcharding** [comment]
Force pushed changes mentioned above.

File: hex/Cargo.toml (hunk 2 above)
Comment (apoelstra) [reply hex/Cargo.toml]: I don't understand what's wrong with `bitcoin-hex` as a name for a hex project whose distinguishing feature is that it is mainained by the rust-bitcoin people, especially when multiple people are using our hex code specifically for that reason.

**Review by apoelstra — APPROVED** [review]
ACK 8de845bc54052acf416f0609e5333105c7fabd1b

File: hex/Cargo.toml (hunk 2 above)
Comment (Kixunil) [reply hex/Cargo.toml]: Projects don't tend to be named after who maintains them. Ideally when one should get a rough idea what a crate is about by just reading the name (just like variables, types...)

There's also a marketing issue: if one used it in a non-bitcoin project it'd cause anyone unfamiliar looking at compilation output to say "why is this pulling in some unrelated bitcoin garbage?".

Comment (tcharding) [reply hex/Cargo.toml]: I also like with what @jeandudey said someplace about not putting marketing terms in the crate name.

Comment (Kixunil) [reply hex/Cargo.toml]: We could also just use some silly short name like `hexify` or `hexer`.

Comment (Kixunil) [reply hex/Cargo.toml]: We *should* put *truthful* marketing names in the name. Saying `fast` is not truthful because we intentionally skip `unsafe` optimizations (which are possible as I demonstrated). Saying `conservative` is truthful because we have conservative MSRV and review process.

Truthful marketing is a good thing because it helps people who are searching for some solution to their problem identify the solution faster.

**Comment by tcharding** [comment]
Needs#1664

File: hex/Cargo.toml (hunk 2 above)
Comment (apoelstra) [reply hex/Cargo.toml]: > Projects don't tend to be named after who maintains them

They are named after the ecosystem that they exist in.

> There's also a marketing issue: if one used it in a non-bitcoin project it'd cause anyone unfamiliar looking at compilation output to say "why is this pulling in some unrelated bitcoin garbage?".

Lol. If people want to use actual garbage because they don't like our name, they're welcome to.

Comment (jeandudey) [reply hex/Cargo.toml]: >There's also a marketing issue: if one used it in a non-bitcoin project it'd cause anyone unfamiliar looking at compilation output to say "why is this pulling in some unrelated bitcoin garbage?".

I do use `bitcoin_hashes` even though it isn't explicitly related to Bitcoin.

It'd be better to keep a organization-like prefix so people know that if a crate starts with `bitcoin-.*` it is maintained by `rust-bitcoin` org.

Comment (Kixunil) [reply hex/Cargo.toml]: > `bitcoin_hashes`

False analogy, that crate specifically does contain things specific to Bitcoin - namely reversed `sha256d` and `hash160`. (Though I would've liked to have non-specific crate for algos alone.) Hex formatting is a universal concept and it cannot be possibly specific to Bitcoin.

>  crate starts with `bitcoin-.*` it is maintained by `rust-bitcoin` org.

That's very wrong and very dangerous assumption. Anyone can upload a crate starting with `bitcoin-`.

Anyway, if you all want it I won't block it even though I still dislike it.

Comment (tcharding) [reply hex/Cargo.toml]: To me the name it should be is `hex` but we can't because of crates.io rules and everyone knows these rules so I always assumed everyone knew that the added fluff in names was mainly just that, fluff to get a unique name. Hence, I don't really care, hence I think `rust-hex` is perfectly adequate.

**Comment by tcharding** [comment]
Force push includes changes to use `hex = { package = "bitcoin-hex" ...`

**Review by apoelstra — APPROVED** [review]
ACK 74f57bd592acfbe287f0f3e0ae0348c4f985eb67

**Comment by tcharding** [comment]
Rebased and checked on bugs were introduced, please see new PR description for tips on verifying the PR.

**Comment by tcharding** [comment]
Sorry @apoelstra, didn't see you'd just acked - I'm trying to be super careful with this one since its contentious.

File: hex/Cargo.toml (hunk 2 above)
Comment (Kixunil) [reply hex/Cargo.toml]: > While it isn't the best scheme nor the most ergonomic

That's why I'm proposing to base the name on the description of the main goals: to maintain conservative MSRV and careful development. It's the same as with variables: clearer names are better. Consider the situation of someone looking at the dependency list of a crate unrelated to bitcoin. After seeing our crate he thinks:

* `hex-conservative-msrv` - "Ah, this crate tries to maintaing conservative MSRV"
* `bitcoin-hex` - "What does it have to do with bitcoin?" <spends a bunch of time investigating> "Oh, nothing, the crate just wants conservative MSRV"

If the naming doesn't matter why don't we just generate a random string and use that? Or name it `pink-pony`?

Comment (jeandudey) [reply hex/Cargo.toml]: >If the naming doesn't matter why don't we just generate a random string and use that? Or name it pink-pony?

For consistency

**Review by apoelstra — DISMISSED** [review]
ACK 21d6abf9d9f5fd2039be6fd553d26633222b1c08

**Comment by tcharding** [comment]
Rebase only, no other changes.

**Comment by tcharding** [comment]
For the record I"m not going to push for this PR to be included in the 0.30 release. I think it should be but its March already, February was the target release month, if @Kixunil is still unconvinced of the merit of this PR then I'm happy to release `hashes` as is so we can get on with the 0.30 release.

**Comment by apoelstra** [comment]
I think we'll have to drop it. Let's try to do the release this weekend. I'll merge #1533 shortly.

**Comment by apoelstra** [comment]
Also, needs rebase again, sorry.

**Comment by apoelstra** [comment]
It would also make my life slightly easier if we resolved #1230 first, and then every commit here that changes the dependency structure (the first and last ones specifically) updated the included `Cargo.lock` files.

**Comment by tcharding** [comment]
Sure thing, no sweat.

**Comment by tcharding** [comment]
Hey @Kixunil I'm going to bring this back to life after the lock files PR goes in. Do you have any other immediate concerns that you'd like me to address before doing so? I will not take a positive response as meaning you will not uncover issues later, just asking if you have anything in mind already that I can fix up to help this go along easily.

**Comment by Kixunil** [comment]
I'm going to get some more free time soon, so I'll look into it properly. Looking forward to it, actually. :)

**Comment by tcharding** [comment]
> Looking forward to it, actually. :)

Awesome!

**Comment by tcharding** [comment]
I rebased to make your life easier @Kixunil, includes the two patches from https://github.com/rust-bitcoin/rust-bitcoin/pull/1764

**Comment by tcharding** [comment]
Ugh! I can't remember what the CI fail is caused by, something to do with `--locked` and a git dependency (mutagen).

**Comment by tcharding** [comment]
Added update to minimal/latest lock files (and mentioned them in commit log) to the last commit (fixing CI ... hopefully).

**Comment by tcharding** [comment]
Closing in favour of https://github.com/rust-bitcoin/hex-conservative/pull/1
